x-pack/plugins/ingest_manager/server/plugin.ts

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+import { Observable } from 'rxjs';
+import {
+  CoreSetup,
+  CoreStart,
+  Plugin,
+  PluginInitializerContext,
+  ICustomClusterClient,
+} from 'kibana/server';
+import { LicensingPluginSetup, ILicense } from '../../licensing/server';
+import { EncryptedSavedObjectsPluginStart } from '../../encrypted_saved_objects/server';
+import { SecurityPluginSetup } from '../../security/server';
+import { PluginSetupContract as FeaturesPluginSetup } from '../../features/server';
+import { PLUGIN_ID } from './constants';
+import { licenseService, configService, appContextService } from './services';
+import { registerDatasourceRoutes, registerAgentConfigRoutes } from './routes';
+import { IngestManagerConfigType } from '../common';
+
+export interface IngestManagerSetupDeps {
+  licensing: LicensingPluginSetup;
+  security?: SecurityPluginSetup;
+  features?: FeaturesPluginSetup;
+}
+
+export interface IngestManagerAppContext {
+  clusterClient: ICustomClusterClient;
+  encryptedSavedObjects: EncryptedSavedObjectsPluginStart;
+  security?: SecurityPluginSetup;
+}
+
+export class IngestManagerPlugin implements Plugin {
+  private licensing$!: Observable<ILicense>;
+  private config$!: Observable<IngestManagerConfigType>;
+  private clusterClient!: ICustomClusterClient;
+  private security: SecurityPluginSetup | undefined;
+
+  constructor(private readonly initializerContext: PluginInitializerContext) {}
+
+  public async setup(core: CoreSetup, deps: IngestManagerSetupDeps) {
+    this.licensing$ = deps.licensing.license$;
+    this.config$ = this.initializerContext.config.create<IngestManagerConfigType>();
+    this.clusterClient = core.elasticsearch.createClient(PLUGIN_ID);
```

Comment:
If you're not doing anything special with this client, can you use the ones provided by default, either via request context or `core.elasticsearch.dataClient`

Reply:
@jfsiii Let's change this line to `this.clusterClient = core.elasticsearch.dataClient;`
I don't think we'll be doing anything special, but we may need to access the client outside of request context.

Reply:
Would you be willing to use the recommended approach until there is a known situation where you need to access the client outside of request context?  If there is a legit reason to access it outside of request context, that may be something the platform team should be aware of, and then they can expose it on the `CoreStart` contract, instead of having plugins save references to `CoreSetup` functionality as a workaround.

Reply:
I think actually this might cause problems doing it this way because `context.core.elasticsearch.dataClient` needs to be scoped to the user hitting the API.  Storing a reference could end up calling `callAsCurrentUser` with the wrong user.   At least that is my takeaway from this convo https://elastic.slack.com/archives/C5TQ33ND8/p1569345861164200?thread_ts=1569344427.160400&cid=C5TQ33ND8 5 months ago.